From 4f07680df03b196c30d165b46c0822cf92beccea Mon Sep 17 00:00:00 2001
From: Paul Moore <pmoore@redhat.com>
Date: Tue, 6 May 2014 15:00:17 -0400
Subject: [security] selinux: Increase ebitmap_node size for 64-bit configuration

Message-id: <20140506150017.16038.89784.stgit@localhost>
Patchwork-id: 79413
O-Subject: [RHEL7 PATCH 2/2] SELinux: Increase ebitmap_node size for 64-bit configuration
Bugzilla: 922752
RH-Acked-by: Eric Paris <eparis@redhat.com>

From: Waiman Long <Waiman.Long@hp.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087656
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7419167
Upstream: in Linus' tree, see commit ID below
Testing: Tested as part of BZ #922752, see the BZ for details

 commit a767f680e34bf14a36fefbbb6d85783eef99fd57
 Author: Waiman Long <Waiman.Long@hp.com>
 Date:   Tue Jul 23 17:38:41 2013 -0400

 SELinux: Increase ebitmap_node size for 64-bit configuration

 Currently, the ebitmap_node structure has a fixed size of 32 bytes. On
 a 32-bit system, the overhead is 8 bytes, leaving 24 bytes for being
 used as bitmaps. The overhead ratio is 1/4.

 On a 64-bit system, the overhead is 16 bytes. Therefore, only 16 bytes
 are left for bitmap purpose and the overhead ratio is 1/2. With a
 3.8.2 kernel, a boot-up operation will cause the ebitmap_get_bit()
 function to be called about 9 million times. The average number of
 ebitmap_node traversal is about 3.7.

 This patch increases the size of the ebitmap_node structure to 64
 bytes for 64-bit system to keep the overhead ratio at 1/4. This may
 also improve performance a little bit by making node to node traversal
 less frequent (< 2) as more bits are available in each node.

 Signed-off-by: Waiman Long <Waiman.Long@hp.com>
 Acked-by:  Stephen Smalley <sds@tycho.nsa.gov>
 Signed-off-by: Paul Moore <pmoore@redhat.com>
 Signed-off-by: Eric Paris <eparis@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/security/selinux/ss/ebitmap.h b/security/selinux/ss/ebitmap.h
index e7eb3a9..712c8a7 100644
--- a/security/selinux/ss/ebitmap.h
+++ b/security/selinux/ss/ebitmap.h
@@ -16,7 +16,13 @@
 
 #include <net/netlabel.h>
 
-#define EBITMAP_UNIT_NUMS ((32 - sizeof(void *) - sizeof(u32)) \
+#ifdef CONFIG_64BIT
+#define EBITMAP_NODE_SIZE 64
+#else
+#define EBITMAP_NODE_SIZE 32
+#endif
+
+#define EBITMAP_UNIT_NUMS ((EBITMAP_NODE_SIZE-sizeof(void *)-sizeof(u32))\
      / sizeof(unsigned long))
 #define EBITMAP_UNIT_SIZE BITS_PER_LONG
 #define EBITMAP_SIZE  (EBITMAP_UNIT_NUMS * EBITMAP_UNIT_SIZE)
-- 
1.7.1